src/picongpu/include/particles/bremsstrahlung/SynchrotronFunctions.hpp

```diff
@@ -0,0 +1,199 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "types.h"
+#include "cuSTL/container/HostBuffer.hpp"
+#include "cuSTL/container/HostBuffer.hpp"
+#include "cuSTL/cursor/Cursor.hpp"
+#include "cuSTL/cursor/navigator/PlusNavigator.hpp"
+#include "cuSTL/cursor/tools/LinearInterp1D.hpp"
+#include "cuSTL/cursor/BufferCursor.hpp"
+#include "simulation_defines.hpp"
+#include <boost/shared_ptr.hpp>
+#include <boost/array.hpp>
+#include <boost/numeric/odeint.hpp>
+#include <boost/math/special_functions/bessel.hpp>
+#include <cmath>
+
+namespace picongpu
+{
+namespace particles
+{
+namespace bremsstrahlung
+{
+
+namespace detail
+{
+
+/** Map `x` to the internal lookup table and return the result of the
+ * first or the second synchrotron function for `x`.
+ */
+struct MapToLookupTable
+{
+    typedef typename ::PMacc::result_of::Functor<
+        ::PMacc::cursor::tools::LinearInterp1D<float_X>,
+        ::PMacc::cursor::BufferCursor<float_X, DIM1> >::type LinInterpCursor;
+
+    typedef float_X type;
+
+    LinInterpCursor linInterpCursor;
+
+    /** constructor
+     *
+     * @param linInterpCursor lookup table of the first or the second
+     * synchrotron function.
+     */
+    HDINLINE MapToLookupTable(LinInterpCursor linInterpCursor)
+        : linInterpCursor(linInterpCursor) {}
+
+    HDINLINE float_X operator()(const float_X x) const
+    {
+        /* This mapping increases the sample point density for small values of x
+         * where the synchrotron functions have a divergent slope. Without this mapping
+         * the emission probabilty of low-energy photons is underestimated.
+         *
+         * This is the inverse mapping of the mapping in @see:`SynchrotronFunctions::init()`
+         */
+        const float_X x_m = PMacc::algorithms::math::pow(x, float_X(1.0/3.0));
+
+        const float_X cutOff = static_cast<float_X>(SYNC_FUNCS_CUTOFF);
+
+        if(x_m >= cutOff)
+            return float_X(0.0);
+        else
+            return this->linInterpCursor[x_m / static_cast<float_X>(SYNC_FUNCS_STEP_WIDTH)];
+    }
+};
+
+typedef ::PMacc::cursor::Cursor<
+    MapToLookupTable,
+    ::PMacc::cursor::PlusNavigator,
+    float_X> SyncFuncCursor;
+
+} // namespace detail
+
+
+/** Lookup table for synchrotron functions.
+ *
+ * Provides cursors for the first and the second synchrotron function
+ */
+class SynchrotronFunctions
+{
+public:
+    typedef detail::SyncFuncCursor SyncFuncCursor;
+private:
+
+    typedef boost::shared_ptr<PMacc::container::DeviceBuffer<float_X, DIM1> > MyBuf;
+    MyBuf dBuf_SyncFuncs[2];
+
+    struct BesselK
+    {
+        template<typename State, typename Time>
```

Comment:
please follow the [coding guideline](https://github.com/ComputationalRadiationPhysics/picongpu/wiki/Coding-Guide-Lines) for template type names
~~`template<typename State, typename Time>`~~ but `template<typename T_State, typename T_Time>`